drivers/clk/clk.h

```diff
@@ -0,0 +1,399 @@
+ *        clk_disable_unused for gate clocks with special needs.
+ *        Called with enable_lock held.  This function must not
+ *        sleep.
+ *
+ * @recalc_rate: Recalculate the rate of this clock, by querying hardware. The
+ *        parent rate is an input parameter.  It is up to the caller to
+ *        ensure that the prepare_mutex is held across this call. If the
+ *        driver cannot figure out a rate for this clock, it must return
+ *        0. Returns the calculated rate. Optional, but recommended - if
+ *        this op is not set then clock rate will be initialized to 0.
+ *
+ * @round_rate:    Given a target rate as input, returns the closest rate actually
+ *        supported by the clock. The parent rate is an input/output
+ *        parameter.
+ *
+ * @determine_rate: Given a target rate as input, returns the closest rate
+ *        actually supported by the clock, and optionally the parent clock
+ *        that should be used to provide the clock rate.
+ *
+ * @set_parent:    Change the input source of this clock; for clocks with multiple
+ *        possible parents specify a new parent by passing in the index
+ *        as a u8 corresponding to the parent in either the .parent_names
+ *        or .parents arrays.  This function in affect translates an
+ *        array index into the value programmed into the hardware.
+ *        Returns 0 on success, -EERROR otherwise.
+ *
+ * @get_parent:    Queries the hardware to determine the parent of a clock.  The
+ *        return value is a u8 which specifies the index corresponding to
+ *        the parent clock.  This index can be applied to either the
+ *        .parent_names or .parents arrays.  In short, this function
+ *        translates the parent value read from hardware into an array
+ *        index.  Currently only called when the clock is initialized by
+ *        __clk_init.  This callback is mandatory for clocks with
+ *        multiple parents.  It is optional (and unnecessary) for clocks
+ *        with 0 or 1 parents.
+ *
+ * @set_rate:    Change the rate of this clock. The requested rate is specified
+ *        by the second argument, which should typically be the return
+ *        of .round_rate call.  The third argument gives the parent rate
+ *        which is likely helpful for most .set_rate implementation.
+ *        Returns 0 on success, -EERROR otherwise.
+ *
+ * @init:    Perform platform-specific initialization magic.
+ *        This is not used by any of the basic clock types.
+ *        This callback exist for HW which needs to perform some
+ *        initialisation magic for CCF to get an accurate view of the
+ *        clock. It may also be used dynamic resource allocation is
+ *        required. It shall not used to deal with clock parameters,
+ *        such as rate or parents.
+ *        Returns 0 on success, -EERROR otherwise.
+ *
+ */
+struct clk_ops {
+    uint8_t (*get_parent)(const struct clk *clk);
```

Comment:
> If these operations do not vary per clock, get rid of the vector.

These operations vary per type of clock. We need this structure to redirect operation call to the clock-specific implementation instead of writing a bunch of conditional statements...